### ISSUE TITLE
PID units and defaults

### ISSUE THREAD

**Issue by drf5n** [Traumflug/Teacup_Firmware#74]
The default PID parameters are assigned around https://github.com/Traumflug/Teacup_Firmware/blob/master/heater.c#L79  it might be nicer to move these out to config.h, or use some optional overrides like PID_I_PARAMETER, etc., in config.h.

Also, it seems that the qs=250mS sampling period, qC=C/4 temperature resolution and TH_COUNT=8 combine to make the units of the PID parameters be: 

```
P: counts/(PID_SCALE*e_qC)
I: counts/(PID_SCALE*(e_qC)*(qS))=counts/(PID_SCALE*e_qC*qs)
D: counts/(PID_SCALE*(e_qC)/(TH_COUNT*qs))
```

(where e_qC is the error in C/4, TH_COUNTS=8, and counts is the numerator of the PWM output power fraction or duty cycle (as in power=Watts*counts/255))

Interpreting the default {8125,512,24576} PID values in engineering units, these would be kP=32 counts/C, kI=8 counts/(C s), and kD=(194) counts/(C/s).  Alternately,  interpreting them as industrial PID controller or "Standard form" values: kP=32/255/C=12.5%FS/C, Ti=kP/kI=4s, and Td=kD/kP=6.0s.
(Edited to redefine things and move several terms and recalculate the values....)
I'd like to be able to:

1) Set the PID parameters in config.h,
2) Use some sort of consistent unit system. 

The first seems straightforward: add PID_P, PID_I, PID_D, PID_I_LIM to config*.h and modify heater.c to use them if defined.

The second seems harder, in that I'd want the units and values in config.h and M136 report to be consistent with the parameter input commands M130, M131, M132, M133.

Thoughts?

**Comment by Traumflug** [comment]
I just rebased all the branches related to this issue to master. These are the branches *issue74* *issue74-ilimOutput* *issue74units* *ProportionalBandPID*.

@drf5n, do you still remember what the use of all these is, which parts work and which don't? I think it's a good point in time now to integrate them to *experimental*, I just don't know which parts are useful. Apparently they conflict a bit, the approach on *ProportionalBandPID* appears to be independent of the one on *issue74*.

A good target might be to put some PID tuning algorithm into Configtool, operated via M-codes, then to adjust each heater individually by changing parameters in config.h. Ideally in a way understandable by dummies. :-)

Some insight would be very helpful :-)

**Comment by Wurstnase** [comment]
The [defaults](https://github.com/Traumflug/Teacup_Firmware/blob/master/heater.h#L9-L17) are left shifted by 10. 
The inputs are floats? [Reprap-Wiki](http://reprap.org/wiki/G-code#M130:_Set_PID_P_value)
I can't find any conversion.

**Comment by Traumflug** [comment]
Removed branch *issue74units*. The one commit was (except for a comment typo fix) identical to the first commit on branch *issue74*.

**Comment by drf5n** [comment]
Let me look at these again.  After looking at the Marlin Autotune stuff and all its add-on bandaids that override and interfere with the PID,  I think I liked the ProportionalBand idea best--it gives a one-dimensional continuum from hyper-aggressive bang-bang behavior to a slow tapering off.  Moving it into the config tool is a good idea.  It would be nice to compile a firmware with chosen PID parameters already baked in.

UI wise, you could make ProportionalBand a bit more complicated with two adjustment dimensions:  Begin tapering at X degrees from the setpoint, and enter the process time constant to calculate a kI.

**Comment by drf5n** [comment]
Traumflug, could you rebase these again?  I'm terrible at that.

**Comment by Traumflug** [comment]
At your service. Rebased *issue74*, *issue74-ilimOutput* and *pid_autotune* to *experimental*. No guarantees on whether this still compiles, much less on how it works. All rebases went without conflict.

**Comment by drf5n** [comment]
I made some changes to issue74 to make it build from configtool.py on the teensy++2.0 board.   

Configtool doesn't know how to read or set the richer heater table though.

How do you invoke a make -f Makefile-AVR with whatever configtool.py is using?

**Comment by Traumflug** [comment]
Configtool writes config.h and as such, using the Makefile is exactly the same with or without using Configtool. Saving a board overwrites thermistortable.h, so you might want to move your tweaked version away before doing this.

**Comment by drf5n** [comment]
It looks like I need to set MCU on the command line if I'm not using a controller with the Makefile's default MCU.  For board.teensy++v2.0.h to compile from the command line like a configtool build I need:

    MCU=at90usb1286 make -f Makefile-AVR

Otherwise I get some undeclared identifier errors like 'DIO33_WPORT' 

Maybe the board/controller config files should set MCU in addition to CPU?

Also, the configtool build seems to use a verbose, command-echoing build, while the command line build uses some @ prefixes to hide the commands.

**Comment by Traumflug** [comment]
Yes, setting MCU, programmer port and such stuff is expected when using the Makefile. Easiest way is to copy *Makefile-AVR* to *Makefile* for doing this. Then a simple `make` is sufficient for building.

> Maybe the board/controller config files should set MCU in addition to CPU?

Can't work, because *make* doesn't read such source files. *gcc* does, but then it's too late already.
